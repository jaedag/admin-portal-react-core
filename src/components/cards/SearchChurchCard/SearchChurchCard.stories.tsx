import React from 'react'
import { Meta, StoryFn } from '@storybook/react'
import SearchChurchCard, {
  SearchChurchCardPropsInterface,
} from './SearchChurchCard'
import { Church } from '@/types'

export default {
  title: 'Components/Cards/SearchChurchCard',
  component: SearchChurchCard,
  argTypes: {
    church: { control: 'object' },
  },
} as Meta

const Template: StoryFn<SearchChurchCardPropsInterface> = (args) => (
  <SearchChurchCard {...args} />
)

export const WithFellowship = Template.bind({})
WithFellowship.args = {
  church: {
    id: '1',
    name: 'Greater Love',
    levelName: 'Fellowship',
    __typename: 'Fellowship',
    leader: {
      id: '1',
      firstName: 'Dan',
      lastName: 'Abramov',
      nameWithTitle: 'Ps. Dan Abramov',
      pictureUrl: 'https://bit.ly/dan-abramov',
    },
  } as Church,
  onClick: () => alert('Card clicked'),
}

export const WithBacenta = Template.bind({})
WithBacenta.args = {
  church: {
    id: '1',
    name: 'Greater Love',
    levelName: 'Bacenta',
    __typename: 'Bacenta',
    leader: {
      id: '1',
      firstName: 'Dan',
      lastName: 'Abramov',
      nameWithTitle: 'Ps. Dan Abramov',
      pictureUrl: 'https://bit.ly/dan-abramov',
    },
  } as Church,
  onClick: () => alert('Card clicked'),
}
