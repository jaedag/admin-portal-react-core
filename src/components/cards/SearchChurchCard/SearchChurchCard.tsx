import React from 'react'
import {
  Avatar,
  AvatarBadge,
  Box,
  Card,
  CardBody,
  Flex,
  Text,
} from '@chakra-ui/react'
import { getCloudinaryPreset } from '@/utils'
import { Church } from '@jaedag/admin-portal-types'
import { BsPeopleFill } from 'react-icons/bs'
import {
  FaBus,
  FaEye,
  FaHouseUser,
  FaPlaceOfWorship,
  FaSynagogue,
  FaUsers,
} from 'react-icons/fa'

export interface SearchChurchCardPropsInterface {
  church: Church
  onClick?: () => void
}

const SearchIconBadge = ({ church }: { church: Church }) => {
  if (church.__typename === 'Fellowship') {
    return (
      <AvatarBadge
        borderColor="papayawhip"
        bg="tomato"
        boxSize="1.25em"
        padding={1}
      >
        <BsPeopleFill color="white" />
      </AvatarBadge>
    )
  }

  if (church.__typename === 'Bacenta') {
    return (
      <AvatarBadge
        borderColor="blue.200"
        bg="blue.600"
        boxSize="1.25em"
        padding={1}
      >
        <FaBus color="white" />
      </AvatarBadge>
    )
  }

  if (church.__typename === 'Constituency') {
    return (
      <AvatarBadge
        borderColor="yellow.200"
        bg="yellow.600"
        boxSize="1.25em"
        padding={1}
      >
        <FaHouseUser color="white" />
      </AvatarBadge>
    )
  }

  if (church.__typename === 'Council') {
    return (
      <AvatarBadge
        borderColor="purple.200"
        bg="purple.600"
        boxSize="1.25em"
        padding={1}
      >
        <FaUsers color="white" />
      </AvatarBadge>
    )
  }

  if (church.__typename === 'Stream') {
    return (
      <AvatarBadge
        borderColor="pink.200"
        bg="pink.600"
        boxSize="1.25em"
        padding={1}
      >
        <FaSynagogue color="white" />
      </AvatarBadge>
    )
  }

  if (
    church.__typename === 'GatheringService' ||
    church.__typename === 'Campus'
  ) {
    return (
      <AvatarBadge
        borderColor="green.200"
        bg="green.600"
        boxSize="1.25em"
        padding={1}
      >
        <FaPlaceOfWorship color="white" />
      </AvatarBadge>
    )
  }

  if (church.__typename === 'Oversight') {
    return (
      <AvatarBadge
        borderColor="cyan.200"
        bg="cyan.600"
        boxSize="1.25em"
        padding={1}
      >
        <FaEye color="white" />
      </AvatarBadge>
    )
  }

  return <AvatarBadge boxSize="1.25em" bg="white" />
}

const SearchChurchCard = (props: SearchChurchCardPropsInterface) => {
  const { church, onClick } = props

  return (
    <Card
      data-testid="search-church-card"
      direction="row"
      marginY={2}
      onClick={onClick}
    >
      <CardBody>
        <Flex gap="5" alignContent="center" alignItems="center">
          <Avatar
            data-testid="avatar"
            borderRadius={10}
            name={church.name + ' ' + church.levelName}
            size="lg"
            src={getCloudinaryPreset(church.leader?.pictureUrl)}
          >
            <SearchIconBadge church={church} />
          </Avatar>
          <Box>
            <Text marginBottom={0} fontWeight="bold" fontSize="lg">
              {church.name + ' ' + church.levelName}
            </Text>

            {!!church.leader && (
              <Text color="grey" marginTop={0}>
                {church.leader.nameWithTitle}
              </Text>
            )}
          </Box>
        </Flex>
      </CardBody>
    </Card>
  )
}

export default SearchChurchCard
