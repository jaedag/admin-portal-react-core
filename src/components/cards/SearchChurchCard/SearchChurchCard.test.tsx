import React from 'react'
import { render } from '@testing-library/react'
import SearchChurchCard from './SearchChurchCard'
import { Church, Member } from '@jaedag/admin-portal-types'
import '@testing-library/jest-dom'

// Mock church data
const church: Church = {
  id: '1',
  __typename: 'Campus',
  name: 'Greater Love',
  levelName: 'Campus',
  leader: {
    id: '1',
    firstName: 'Dan',
    lastName: 'Abramov',
    nameWithTitle: 'Ps. Dan Abramov',
    pictureUrl: 'https://bit.ly/dan-abramov',
  } as Member,
}

describe('SearchChurchCard', () => {
  it('renders church information correctly', () => {
    const { getByText } = render(<SearchChurchCard church={church} />)

    const nameElement = getByText(`${church.name} ${church.levelName}`)
    expect(nameElement).toBeInTheDocument()

    const leaderElement = getByText(church.leader?.nameWithTitle ?? '')
    expect(leaderElement).toBeInTheDocument()
  })

  it('matches snapshot', () => {
    const { container } = render(<SearchChurchCard church={church} />)
    expect(container).toMatchSnapshot()
  })

  it('does not render picture if church has no picture', () => {
    const churchWithoutPicture = { ...church, pictureUrl: '' }
    const { queryByTestId } = render(
      <SearchChurchCard church={churchWithoutPicture} />
    )

    const pictureElement = queryByTestId('church-picture')
    expect(pictureElement).toBeNull()
  })

  it('calls onClick when clicked', () => {
    const onClick = jest.fn()
    const { getByTestId } = render(
      <SearchChurchCard church={church} onClick={onClick} />
    )

    const card = getByTestId('search-church-card')
    card.click()

    expect(onClick).toHaveBeenCalledTimes(1)
  })
})
