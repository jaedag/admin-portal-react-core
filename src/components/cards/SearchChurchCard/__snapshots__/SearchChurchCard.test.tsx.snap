// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchChurchCard matches snapshot 1`] = `
<div>
  <div
    class="chakra-card css-lgv4hl"
    data-testid="search-church-card"
  >
    <div
      class="chakra-card__body css-0"
    >
      <div
        class="css-hsgtjs"
      >
        <span
          class="chakra-avatar css-6zzhhc"
          data-testid="avatar"
        >
          <div
            aria-label="Greater Love MegaBacenta"
            class="chakra-avatar__initials css-0"
            role="img"
          >
            GM
          </div>
          <div>
            SearchIconBadge
          </div>
        </span>
        <div
          class="css-0"
        >
          <p
            class="chakra-text css-t8vhkl"
          >
            Greater Love MegaBacenta
          </p>
          <p
            class="chakra-text css-1hxe9sv"
          >
            Ps. Dan Abramov
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
